### PR TITLE
[5.2] Fix filtering exclude and include articles

### PR DESCRIPTION
**Pull request by rdeutz**
### Summary of Changes
The new article model has a article id filtering, it was a simple text field where you could add ids. This is changed to a subform with a article modal to chose articles. Further more it is not possible to exclude and include articles by id at the same time. 

I have changed the configuration of the module so that is only allow what is possible.

#### Exclude Mode  
02A9B0B4-0C64-41F5-B220-08A8EF8F9B00
#### Include Mode
4317BFBF-3F9F-42AE-8707-688B8A3AAC8C

### Testing Instructions
Try to exclude some articles and include some articles. Check if the module show the expected articles. 

### Expected result AFTER applying this Pull Request
Module shows the expected articles

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: language/en-GB/mod_articles.ini [hunk 1]
```diff
@@ -51,6 +52,7 @@ MOD_ARTICLES_FIELD_MONTHYEARFORMAT_DESC="Please enter in a valid date format. Se
 MOD_ARTICLES_FIELD_MONTHYEARFORMAT_LABEL="Month and Year Display Format"
 MOD_ARTICLES_FIELD_ONLYARCHIVED_LABEL="Archived Articles"
 MOD_ARTICLES_FIELD_RELATIVEDATE_LABEL="Relative Date"
+MOD_ARTICLES_FIELD_SELECT_ARTICLE_LABEL="Select article"
```
Comment (brianteeman) [inline language/en-GB/mod_articles.ini]:
```suggestion
MOD_ARTICLES_FIELD_SELECT_ARTICLE_LABEL="Select Article"
```

File: language/en-GB/mod_articles.ini [hunk 2]
```diff
@@ -25,7 +25,8 @@ MOD_ARTICLES_FIELD_DATEGROUPINGFIELD_DESC="Select which date field you want the
 MOD_ARTICLES_FIELD_DATEGROUPINGFIELD_LABEL="Date Grouping Field"
 MOD_ARTICLES_FIELD_DATERANGEFIELD_LABEL="Date Range Field"
 MOD_ARTICLES_FIELD_ENDDATE_LABEL="To Date"
-MOD_ARTICLES_FIELD_EXCLUDEDARTICLES_LABEL="Article IDs to Exclude"
+MOD_ARTICLES_FIELD_EX_OR_INCLUDE_LABEL="Exclude or include articles"
```
Comment (brianteeman) [inline language/en-GB/mod_articles.ini]:
```suggestion
MOD_ARTICLES_FIELD_EX_OR_INCLUDE_LABEL="Exclude or Include Articles"
```

File: modules/mod_articles/mod_articles.xml [hunk 3]
```diff
@@ -77,33 +78,72 @@
 					showon="show_child_category_articles:1"
 				/>
 
+				<field
+					name="ex_or_include_articles"
+					type="list"
+					label="MOD_ARTICLES_FIELD_EX_OR_INCLUDE_LABEL"
+					default="0"
+					filter="integer"
+				>
```
Comment (brianteeman) [inline modules/mod_articles/mod_articles.xml:87]:
```suggestion
					>
```

File: modules/mod_articles/mod_articles.xml [hunk 4]
```diff
@@ -77,33 +78,72 @@
...
 				<field
 					name="exclude_current"
 					type="radio"
 					label="MOD_ARTICLES_FIELD_EXCLUDE_CURRENT_LABEL"
 					layout="joomla.form.field.radio.switcher"
 					default="1"
 					filter="integer"
-					>
+					showon="ex_or_include_articles:0"
+				>
```
Comment (brianteeman) [inline modules/mod_articles/mod_articles.xml]:
```suggestion
					>
```

File: modules/mod_articles/mod_articles.xml [hunk 5]
```diff
@@ -77,33 +78,72 @@
...
-					type="textarea"
+					type="subform"
 					label="MOD_ARTICLES_FIELD_EXCLUDEDARTICLES_LABEL"
-					cols="10"
-					rows="3"
-				/>
+					multiple="true"
+					buttons="add,remove"
+					showon="ex_or_include_articles:0"
+				>
```
Comment (brianteeman) [inline modules/mod_articles/mod_articles.xml]:
```suggestion
					>
```

File: modules/mod_articles/mod_articles.xml [hunk 6]
```diff
@@ -77,33 +78,72 @@
...
-					type="textarea"
+					type="subform"
 					label="MOD_ARTICLES_FIELD_INCLUDEDARTICLES_LABEL"
-					cols="10"
-					rows="3"
-				/>
+					multiple="true"
+					buttons="add,remove"
+					showon="ex_or_include_articles:1"
+				>
```
Comment (brianteeman) [inline modules/mod_articles/mod_articles.xml]:
```suggestion
					>
```

**Comment by drmenzelit** [comment]
I got an error testing the PR:
grafik

File: modules/mod_articles/mod_articles.xml [hunk 7]
```diff
@@ -22,7 +22,8 @@
 	<help key="Site_Modules:_Articles" />
 	<config>
 		<fields name="params">
-			<fieldset name="basic">
+			<fieldset name="basic"
+					  addfieldprefix="Joomla\Component\Content\Administrator\Field" >
```
Comment (richard67) [inline modules/mod_articles/mod_articles.xml]:
```suggestion
					addfieldprefix="Joomla\Component\Content\Administrator\Field" >
```
We should not use a mix of tabs and spaces for indentation in XML files.

**Comment by Kostelano** [comment]
The module parameters are also used in the plugin with demo data. We probably need to make changes in file [plugins/sampledata/blog/src/Extension/Blog.php](https://github.com/joomla/joomla-cms/pull/43738/files#diff-53875eb03959b47790d84d0502049405b013a41ffcedce7163943d75f07c622a), because demo data creates 3 modules of type ARTICLES after installation.

**Comment by RickR2H** [comment]
I have tested this item :white_check_mark: successfully on df87c15e7c20a77d8e12910b5c928c67b5d33c11<br /><br />Works! The tabs ans spaces issue has be resolved but that has no influence on the test results. please restore test results when the spacing issue is solved.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44142">issues.joomla.org/tracker/joomla-cms/44142</a>.</sub>

**Comment by RickR2H** [comment]
I have tested this item :white_check_mark: successfully on efdf4bcefe83e24b645558e84e1c52a6a8bdae6a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44142">issues.joomla.org/tracker/joomla-cms/44142</a>.</sub>

**Comment by drmenzelit** [comment]
I have tested this item :white_check_mark: successfully on efdf4bcefe83e24b645558e84e1c52a6a8bdae6a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44142">issues.joomla.org/tracker/joomla-cms/44142</a>.</sub>

**Comment by RickR2H** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44142">issues.joomla.org/tracker/joomla-cms/44142</a>.</sub>

**Comment by brianteeman** [comment]
I do not think it is correct that the option to exclude the current article is **only** available if you have defined a list of articles to exclude. This option should be independent of any other settings.

USE CASE
I want to include articles 1,2,3,4,5 in the module but I do not want to include the article I am currently viewing

**Comment by RickR2H** [comment]
@brianteeman Agreed, but we are in a hurry to get everything done so we have a least a stable module. This behaviour can be fixed in a new PR. I'll put it on the list of things to do regarding the module.

**Comment by brianteeman** [comment]
Less haste more speed. No reason to accept "do it later" approach ever

**Comment by Hackwar** [comment]
Thank you @rdeutz for your contribution.

**Comment by brianteeman** [comment]
grrh

**Comment by rdeutz** [comment]
> I do not think it is correct that the option to exclude the current article is only available if you have defined a list of articles to exclude. This option should be independent of any other settings.
USE CASE
I want to include articles 1,2,3,4,5 in the module but I do not want to include the article I am currently viewing

As I wrote:
> Further more it is not possible to exclude and include articles by id at the same time.

You can only select true or false for 

> $articles->setState('filter.article_id.include', [true|false]);

If you really want to implement the use case you would have to look, if the current article is part of the list of included articles. Doable but I didn't thought so far. We can do this with 5.3.

**Comment by brianteeman** [comment]
That's a disappointing loss of functionality

**Pull request merged**